%md
# Final Project - Flight Delays

#### Team members:
- Carla Cortez
- Redwan Hussain
- Anqi Liu
- Murray Stokely

### 1. Abstract
In the aviation industry, airline companies have taken a keen interest in predicting flight delays because of their financial impact and to retain customer satisfaction. The ability to accurately predict a delay will enable flight operations to proactively mitigate any costs associated with rerouting flights. Our team's goal is to analyze historical flight and weather data from 2015-2021 and build a classification model to predict such delays within a 2-hour window. The objective of phase 2 of this project is to develop our base model (using logistic regression) and evaluate against precision and recall.

This phase consisted of the following steps: Exploratory Data Analysis (EDA), data preprocessing, joins, splitting into train and test sets, model creation, and last, model evaluation. During EDA, we noticed a seasonal pattern in delayed flights and determined which features would be most relevant for experimentation. The final dataset was created by a series of joins that accounted for weather conditions associated with the departure location for each flight. We excluded canceled flights and 2020 data due to expected anomalies from the COVID-19 pandemic.

Our joined data set has access to 63 features to predict DEP_DEL15, the boolean indicator for flight delay.  We have also extracted features from the METAR SkyObservation text for clear skies and overcast skies to identify potentially useful features for Phase 3.  Our initial logistic regression models so far incorporate only a small number of features and predict that all flights are not delayed.  This gives us results equivalent to our baseline model.  In Phase 3 we will continue to improve our logistic regression model by adding additional features from the joined data set.

### 2. Data description

This project relies on three main datasets.

1. **Flights data** This is a subset of the passenger flight's on-time performance data taken from the TranStats data collection available from the U.S. Department of Transportation (DOT).  There are approximately 100 features in this data set.

2. **Weather** This provides weather information from NOAA for the same time period 2015-2021 as the flights data.

3. **Airports** This provides more detailed information about the airports in the flights dataset.

#### 2.1 EDA - Pre Join

In [ ]:
import numpy as np
from matplotlib import pyplot as plt
import pandas as pd
import seaborn as sns  
sns.set(style="darkgrid")  
from pyspark.sql.functions import col,isnan,when,count
import statsmodels.api as sm
from tabulate import tabulate
from pyspark.sql.functions import to_timestamp
from pyspark.sql.functions import to_utc_timestamp
data_BASE_DIR = "dbfs:/mnt/mids-w261/datasets_final_project_2022/"

In [ ]:
blob_container = "w261container" # The name of your container created in https://portal.azure.com
storage_account = "mstokely" # The name of your Storage account created in https://portal.azure.com
secret_scope = "w261t21scope" # The name of the scope created in your local computer using the Databricks CLI
secret_key = "saskey" # The name of the secret key created in your local computer using the Databricks CLI
blob_url = f"wasbs://{blob_container}@{storage_account}.blob.core.windows.net"
mount_path = "/mnt/mids-w261"

spark.conf.set(
  f"fs.azure.sas.{blob_container}.{storage_account}.blob.core.windows.net",
  dbutils.secrets.get(scope = secret_scope, key = secret_key)
)

##### 2.1.1 Full Flight (Main) Data

In [ ]:
df_airlines_full = spark.read.parquet(f"{data_BASE_DIR}parquet_airlines_data/")
df_airlines_full.createOrReplaceTempView("df_airlines_full_tb")

__Flight Data - Response Variable Check__


Please refer to the ___Flight delayed pct by year___ and ___Flight delayed cnt___ tab below for our response variable check.


___Flight delayed pct by year___
<br>`The plot shows the delayed percentage out of all non-cancelled flights over month by year`
<br>All years except for the 2020 are showing similar sesonality trends.


___Flight delayed cnt___
<br>`The plot shows the total delayed flights by year - month`
<br>The number of flights dropped significantly starting from 2020 and is slowing recovring.

In [ ]:
%sql

select
year, month,
concat(cast(year as string), case when month >= 10 then cast(month as string) else concat('0',cast(month as string)) end) as year_month,
sum(case when dep_delay>15 then 1 else 0 end) as delay_cnt,
sum(case when dep_delay>15 then 1 else 0 end)/count(dep_delay) as delay_pct
from df_airlines_full_tb
group by year, month, year_month
order by year, month;

year month year_month delay_cnt delay_pct 2015 1 201501 175414 0.19137005221345332 2015 2 201502 176200 0.21533392645894234 2015 3 201503 183702 0.18605635714329122 2015 4 201504 153274 0.15936227640788692 2015 5 201505 173236 0.17617566718260902 2015 6 201506 226180 0.22837283597099348 2015 7 201507 212222 0.20555741741264716 2015 8 201508 185980 0.18384119157719095 2015 9 201509 112126 0.121081685809191 2015 10 201510 115470 0.11933008974300678 2015 11 201511 134008 0.14451947984720673 2015 12 201512 189304 0.20072484206374283 2016 1 201601 136416 0.15703320333184453 2016 2 201602 119452 0.1431193627850358 2016 3 201603 154850 0.1631356601805712 2016 4 201604 126504 0.13821525498542503 2016 5 201605 143744 0.15067031435070177 2016 6 201606 191952 0.19875375086199934 2016 7 201607 221310 0.22417631840715388 2016 8 201608 196606 0.19989995139915528 2016 9 201609 116538 0.12850177197435655 2016 10 201610 121118 0.12935811171633024 2016 11 201611 113746 0.12646875694907717 2016 12 201612 196140 0.2161455693133681 2017 1 201701 180110 0.20398617365383395 2017 2 201702 119628 0.14793178923417463 2017 3 201703 164338 0.17110663631952788 2017 4 201704 173482 0.18810897790384737 2017 5 201705 183310 0.18991249790206313 2017 6 201706 212336 0.217037021894229 2017 7 201707 213046 0.21150035540271694 2017 8 201708 198884 0.1989998178940143 2017 9 201709 113238 0.12766921769090966 2017 10 201710 131800 0.13826963628161684 2017 11 201711 102254 0.11290675205653398 2017 12 201712 165506 0.18026501809119883 2018 1 201801 191756 0.17344223446304688 2018 2 201802 179308 0.17523244382159728 2018 3 201803 188942 0.1589535124366725 2018 4 201804 192398 0.16320710382420953 2018 5 201805 225536 0.18512869931525627 2018 6 201806 255276 0.2074739921976593 2018 7 201807 271172 0.2138088074355117 2018 8 201808 266846 0.21100550354251013 2018 9 201809 171730 0.14858724016917124 2018 10 201810 179972 0.1470763216046235 2018 11 201811 200750 0.17299919855912996 2018 12 201812 201958 0.17193299688753583 2019 1 201901 191262 0.16847418212568047 2019 2 201902 215728 0.20809186106631067 2019 3 201903 205050 0.16542987564320394 2019 4 201904 210720 0.17622294812160988 2019 5 201905 241246 0.19333490408025555 2019 6 201906 290754 0.23295990821145338 2019 7 201907 263912 0.20400887425307082 2019 8 201908 255232 0.19700546019552964 2019 9 201909 157136 0.13178172352138048 2019 10 201910 195222 0.15466803993028047 2019 11 201911 162346 0.13568952202010612 2019 12 201912 247574 0.19957501213214607 2020 1 202001 78854 0.13128176782702652 2020 2 202002 79398 0.13946184638786477 2020 3 202003 48898 0.09084237923878775 2020 4 202004 8657 0.047207211137346425 2020 5 202005 7904 0.04676175995551007 2020 6 202006 13867 0.06220142910328927 2020 7 202007 25119 0.07174540791116012 2020 8 202008 24320 0.06522467917342738 2020 9 202009 18117 0.05633658286300667 2020 10 202010 23834 0.0680401835035242 2020 11 202011 21695 0.059865451053543635 2020 12 202012 35422 0.09635519382839298 2021 1 202101 29397 0.08213907508668793 2021 2 202102 35749 0.11377495162440167 2021 3 202103 42444 0.09673317015137634 2021 4 202104 46006 0.10263262979661313 2021 5 202105 65213 0.13217868942173217 2021 6 202106 125757 0.23394692537368963 2021 7 202107 141044 0.24554073486043362 2021 8 202108 124851 0.22200982632337957 2021 9 202109 72698 0.13677686883358295 2021 10 202110 101340 0.18318938246342178 2021 11 202111 83826 0.15401745112197596 2021 12 202112 119536 0.22183703692706264

Output can only be rendered in Databricks

Output can only be rendered in Databricks

##### 2.1.2 Data To Be Excluded

1. Flights from 2020 is showing very different trends comparing to all the other years, which is considered to be an outlier year and will be excluded from the further analysis.
2. Flight cancelled are not included in the "delay" analysis.

##### 2.1.3 Variables to be investigated

##### 2.1.3.1 Stations

With Airport Codes to Connect with All Other Tables

In [ ]:
df_airport_codes = spark.read.option("header",True).csv(f"{blob_url}/airports_tz.csv")
select_columns = ['IATA', 'ICAO', 'tz_time_zone']
df_iata_icao_codes = df_airport_codes.select(select_columns).filter(col("ICAO").isNotNull() & col("IATA").isNotNull())

df_stations = spark.read.parquet(f"{data_BASE_DIR}stations_data/*")
df_unique_neighbor_stations = df_stations.filter(col('neighbor_id') == col('station_id'))
df_neighbor_stations = df_unique_neighbor_stations.join(df_iata_icao_codes, df_unique_neighbor_stations.neighbor_call ==  df_iata_icao_codes.ICAO, "inner")
select_station_cols = [
  'station_id',
  'IATA',
  'tz_time_zone'
]
df_neighbor_stations = df_neighbor_stations.select(select_station_cols)
df_neighbor_stations.createOrReplaceTempView("df_neighbor_stations_temp")

In [ ]:
# df_neighbor_stations.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in df_neighbor_stations.columns]
#    ).toPandas()

##### 2.1.3.2 Flight

In [ ]:
df_airlines = spark.read.parquet(f"{data_BASE_DIR}parquet_airlines_data/")

#finding and removing duplicates
df_airlines_distinct = df_airlines.distinct()
# print("Total: ", str(df_airlines.count()), " Distinct: ", str(df_airlines_distinct.count()), " Removed duplicates: "+str(df_airlines.count() - df_airlines_distinct.count()))

##Filter cancelled flights 
df_airlines_clean =  df_airlines_distinct.filter(col('CANCELLED') == 0)

#Create new df_flights dataframe with only the required columns
select_airline_cols = ['QUARTER',
 'MONTH',
 'DAY_OF_MONTH',
 'DAY_OF_WEEK',
 'FL_DATE',
 'OP_UNIQUE_CARRIER',
 'TAIL_NUM',
 'OP_CARRIER_FL_NUM',
 'ORIGIN',
 'ORIGIN_CITY_NAME',
 'ORIGIN_STATE_ABR',
 'ORIGIN_WAC',
 'DEST',
 'DEST_CITY_NAME',
 'DEST_STATE_ABR', 
 'DEST_WAC',
 'CRS_DEP_TIME',
 'DEP_DEL15',
 'DEP_DELAY_GROUP',
 'DEP_TIME_BLK',
 'TAXI_IN',
 'CRS_ARR_TIME',
 'CRS_ELAPSED_TIME',
 'DISTANCE',
 'DISTANCE_GROUP',
 'YEAR']
df_airlines_clean =  df_airlines_clean.select(select_airline_cols)
df_airlines_clean.createOrReplaceTempView("df_airlines_clean_temp")

__2.1.3.2.1 Flight - Departure and Arrival Time__

In [ ]:
%sql

select
 CRS_DEP_TIME
from
  df_airlines_clean_temp

CRS_DEP_TIME 620 620 2155 939 800 1435 1406 959 959 1153 625 625 1615 954 900 845 1051 1555 730 1406 1225 2000 711 1739 1539 1751 1751 1124 1046 1645 720 1146 2125 1349 1540 1715 1623 815 630 1010 700 700 610 1736 1205 1739 840 1545 715 2104 1805 1330 1330 1115 1235 1519 1345 1945 1152 1152 1500 1258 605 1352 850 2020 1443 915 1515 955 1705 2000 1415 1304 1020 1930 1710 1041 800 1430 1338 830 620 1956 1100 1410 1355 1105 800 800 730 1350 2025 950 1750 1128 930 1408 1932 835 700 816 640 2149 1225 1555 1630 125 1520 1330 1840 515 735 750 756 1307 720 705 840 800 1155 1742 704 1255 1300 1428 1335 1259 1740 1845 2045 2005 1326 1520 1439 1710 1103 1550 1754 1925 1720 945 1727 2000 1659 1948 2355 2045 700 1050 1845 1729 2225 1115 2315 1402 2043 1852 620 600 1409 1645 920 600 1740 718 1000 1045 1045 1613 600 810 1339 1855 1311 840 730 824 1900 1745 830 923 1335 1100 2359 840 1011 2342 1601 600 1056 1346 1930 924 730 645 810 910 1537 1204 2017 1433 1530 1640 2250 1930 1050 949 1210 1615 600 1321 2355 910 710 840 945 1121 1459 1653 600 1030 2023 1920 632 1103 1750 535 1815 850 645 2034 1822 820 1420 743 2049 1401 1123 1300 1354 1739 1630 815 940 630 1854 855 540 1540 821 1905 730 930 2359 730 110 1916 1900 1945 1230 730 905 1602 1021 1040 2125 1615 1430 1710 1135 503 815 1310 1912 952 1345 1520 1925 530 630 1951 1625 1509 1419 700 1817 915 1300 1350 1800 1845 1625 1105 1449 750 500 1520 1130 1840 1915 810 1110 2217 730 2000 1350 1050 1430 1445 1330 1855 1600 1550 1215 700 1220 1110 1515 1620 1600 1539 1536 1500 730 1540 1530 815 645 1520 750 1815 1445 1345 2020 1959 2142 1240 700 2015 1135 940 810 526 1925 1105 1127 2140 745 1615 920 700 2155 1205 1530 1926 624 545 2005 742 742 700 1945 730 1100 1100 1530 2035 1900 1945 825 2306 1940 1015 1100 606 1005 530 1045 240 1930 658 2019 1335 1820 1200 1040 1653 1235 1230 2005 1259 1545 1800 1300 1208 1157 1715 1000 700 545 1535 1551 1555 1555 2025 926 45 645 1532 2055 1505 1525 1815 1710 600 1656 1205 1208 1025 1345 1657 2205 2208 1000 830 1840 1029 1454 1730 1730 1625 1025 1827 810 1847 1351 950 1542 2110 2210 1345 1145 1454 801 941 700 1149 1300 1957 613 1538 1655 600 1942 2145 710 615 914 1713 1155 2110 1430 1830 1008 503 1703 1703 2230 1325 1015 613 955 2018 1930 818 1145 930 940 1535 1420 1617 1558 1600 1834 1905 1100 1510 1325 1200 630 2200 2000 2015 1645 830 1417 2345 1013 1731 1027 1027 710 1824 1255 1651 2020 1017 1205 1205 931 1150 1825 1012 1910 637 1441 1053 705 815 629 632 2016 1505 1004 1110 1505 1510 1715 1322 839 1715 1147 1025 1907 1530 600 730 1018 1512 1128 1041 1120 730 600 2141 1437 739 1304 1145 1506 1839 1825 659 600 1007 855 1005 1905 847 805 1047 1617 1030 1705 1645 1010 1610 715 1431 1842 750 1030 1659 1831 1441 1802 1838 1800 1810 1440 1720 633 1037 733 1412 1500 810 611 655 1620 700 1625 615 2359 2000 1215 1707 1045 930 530 1619 1619 1449 2236 1454 600 610 1610 1456 848 620 620 1900 1732 1451 929 1723 1723 1558 1711 1022 1305 1400 1940 1455 1306 1400 1440 1103 755 2259 610 925 1307 545 1631 1845 2037 2150 1916 1529 1608 2254 1850 832 1941 1956 1946 701 905 909 1430 1320 1950 1135 2151 2237 1205 1240 2151 905 900 1456 1011 609 757 1822 1300 2020 1213 1625 635 1709 1256 838 928 645 705 1755 550 715 600 1009 830 830 2320 810 2030 1728 1734 1725 1701 1419 855 1015 1015 1441 1325 1220 2359 2109 2327 1339 1023 1056 600 936 1018 1018 1339 2110 1412 1050 1940 1008 1527 2103 1826 1920 1824 625 2043 1855 610 923 2105 2330 2307 1959 1539 959 1620 1240 725 645 600 1606 1000 1537 853 2051 2051 920 1155 2055 1155 1510 2230 1935 1415 1355 1728 715 2005 955 645 720 1030 720 600 1630 1635 2003 1325 1400 1735 2315 2310 1655 1153 832 1730 1815 1230 1705 630 1306 820 1646 1450 700 1205 640 1055 1735 1232 2200 1412 1251 1540 1430 615 1355 1155 1702 905 850 1635 1830 1845 1540 1517 1015 1445 1105 1410 1805 1258 915 1817 1000 2010 740 1615 700 2205 1612 1555 1450 1810 1450 535 1311 1732 830 2115 1442 1446 1305 1540 1454 750 1610 1050 1130 1450 2045 1719 1255 1920 1250 1445 740 650 809 103

Output can only be rendered in Databricks

In [ ]:
%sql

select
 CRS_ARR_TIME
from
  df_airlines_clean_temp

CRS_ARR_TIME 710 1215 1910 1730 1335 2156 1950 2311 1335 1945 1820 744 1055 1523 1400 1425 1442 725 2300 1050 1755 1416 1440 614 1719 1828 1645 15 1140 1705 2035 1735 1310 2325 2339 637 2134 1055 1750 2321 1839 1945 945 2300 2244 1840 2145 717 903 2135 927 801 1408 1419 942 1325 1319 2345 1305 2220 2116 1920 700 1448 2130 735 2357 2359 2310 1040 1040 1040 1528 908 1719 1830 1148 1313 2225 1606 2138 937 1557 733 814 1015 746 2149 2241 2241 1403 2047 927 1943 1918 2101 855 2114 1559 900 749 1940 645 1549 2047 1354 1545 1545 1800 1306 811 2106 1726 1927 1020 2249 1450 1350 2050 2309 2057 844 1452 2142 2313 2157 2132 2114 2058 1533 1533 919 1300 1653 2200 1705 1320 1017 1718 2247 835 14 1310 2258 1031 759 2127 2038 845 1836 1341 1618 825 1305 1250 1402 1836 1640 900 1226 1919 2130 1531 1645 1731 1844 1449 2326 1241 1054 1123 1938 2347 954 2344 2044 2201 1829 1938 2036 2014 1955 1352 1458 2137 650 2226 1412 1649 1043 1145 1952 1209 1029 820 1542 2004 1522 1930 2146 2138 1448 1632 1056 1647 1556 1520 1449 1105 2300 1908 1939 1405 1627 604 1116 1354 1910 1155 47 1125 2112 2058 1734 1734 1542 1310 2129 2305 2039 2053 1227 929 754 1542 2146 2136 2104 2314 2227 1535 752 2047 1809 1559 1350 1021 25 1730 1749 1526 2203 930 1455 1455 1455 1046 1940 2200 900 13 2012 1020 645 1921 1745 1739 1804 1900 1244 1917 1200 1350 1340 1555 2028 2222 1738 1110 1005 1100 1113 2230 742 1021 1015 1739 1301 2012 1941 2252 910 902 805 535 2312 1635 1520 1730 1049 846 1124 1111 1015 1444 220 2122 1600 1938 1245 1020 1849 2000 1931 610 720 1420 2057 1245 935 954 1805 1052 1755 1735 2212 1430 1719 1614 1243 1639 1820 2028 1742 55 2257 1021 1106 1636 1745 1316 707 2029 443 1918 924 914 1316 1423 1020 1354 2130 1443 8 1927 1214 1215 809 1527 809 840 1739 1608 1614 1511 809 1440 2038 1317 1103 1356 1010 2204 1355 514 937 2153 2354 2354 2248 2303 1240 755 1455 1559 1100 2325 1005 1814 1632 1101 2328 1049 1259 2017 1801 2105 917 1310 949 1115 1606 1245 1843 1130 1542 2135 1433 1814 1600 1300 1038 1006 2134 732 1950 1059 2012 2001 1817 1855 2148 1525 2010 1013 2138 1820 937 1035 1635 1009 1115 1639 1852 1218 1935 1645 2015 2317 2310 2359 2119 1030 2215 853 1751 505 2312 615 631 905 1345 1950 945 1044 1530 1223 935 2310 1612 1258 1846 23 1420 1700 2106 1238 1822 2155 933 1459 2010 1712 2247 1227 1139 1940 1314 2123 915 1230 2144 931 1504 1257 733 1436 942 1235 1607 2134 2010 1441 1240 1055 1259 2329 1055 2203 1537 1350 1344 917 1347 841 841 1549 939 1240 648 1917 2020 1220 1718 911 1730 1236 2210 1137 909 2200 906 2030 1843 2059 1647 1345 1011 2117 1659 1901 914 915 906 1534 1534 1420 1534 1925 1628 1209 831 1610 909 946 1405 1500 836 2256 2017 1956 847 1351 1639 846 1716 1512 1521 1719 1624 1930 1610 2144 1715 2005 1821 1331 1605 1929 2357 1334 1238 1825 749 746 1437 2158 1613 925 1807 720 648 2244 2359 848 1135 937 1149 1132 1040 1629 1408 2205 1514 1105 1915 900 1220 133 2202 2145 2145 2043 940 1727 2248 1058 1357 1807 2110 600 1438 1226 2030 648 1239 1215 1404 2350 1653 1316 1450 1452 2030 2129 605 559 1115 617 1422 935 1259 842 2059 910 1232 1624 1646 1053 909 2326 2037 2242 113 1212 2017 849 2332 1221 2123 705 705 1202 917 2314 951 1914 2307 1810 1625 1835 828 705 1235 1733 1420 55 1720 1500 710 1830 1439 1020 1140 2145 1600 1104 935 2253 910 1056 1433 1745 1647 2159 1755 1910 1203 1120 1455 150 938 940 1624 1904 1117 1045 1950 650 2224 2120 1724 1225 810 1735 1555 1605 1650 1915 1030 1855 2225 755 1623 1646 1643 1855 1625 1130 2219 2048 1430 1629 2105 2325 1910 1949 2130 1415 1905 1040 2225 955 2101 1434 725 631 2200 1850 15 1520 145 2215 1140 1535 1353 1615 1355 2044 2310 2120 720 715 1440 1620 2050 1940 1945 1625 751 2353 910 1115 1015 835 2310 1250 1805 1302 1859 925 1905 1121 2137 939 1919 1705 650 910 2055 2336 2249 1729 929 1110 2220 745 1315 1848 1418 1433 1420 55 1031 2120 1940 1927 1335 1444 1000 900 740 1258 1935 1446 1523 1435 1949 1142 1127 2115 1802 1030 1820 900 1500 1527 2155 1050 1635 1815 1549 2155 1520 1627 5 2035 1645 1030 1322 1235 855 1335

Output can only be rendered in Databricks

In [ ]:
%sql

select
 CRS_DEP_TIME,
 CRS_ARR_TIME,
 CRS_ARR_TIME - CRS_DEP_TIME,
 CRS_ELAPSED_TIME
from
  df_airlines_clean_temp
where CRS_ARR_TIME < CRS_DEP_TIME

CRS_DEP_TIME CRS_ARR_TIME (CRS_ARR_TIME - CRS_DEP_TIME) CRS_ELAPSED_TIME 2250 15 -2235 85.0 841 835 -6 54.0 1610 14 -1596 304.0 2220 604 -1616 284.0 2220 47 -2173 147.0 2035 25 -2010 170.0 2315 13 -2302 58.0 2350 535 -1815 225.0 2155 220 -1935 205.0 1635 55 -1580 320.0 2359 443 -1916 224.0 2245 8 -2237 83.0 2355 809 -1546 314.0 2355 809 -1546 314.0 2350 514 -1836 204.0 2145 505 -1640 380.0 2230 23 -2207 293.0 2250 720 -1530 330.0 2240 648 -1592 308.0 2245 133 -2112 288.0 2140 605 -1535 325.0 2204 559 -1645 295.0 2300 617 -1683 257.0 1950 113 -1837 203.0 2326 55 -2271 149.0 1800 150 -1650 290.0 2240 15 -2225 95.0 1825 145 -1680 260.0 2230 55 -2175 85.0 2150 5 -2145 135.0 2300 640 -1660 280.0 2145 5 -2140 80.0 1830 20 -1810 230.0 2315 733 -1582 318.0 1935 55 -1880 200.0 2155 140 -2015 165.0 2357 832 -1525 335.0 2145 25 -2120 220.0 2250 732 -1518 342.0 2150 5 -2145 75.0 2340 742 -1598 302.0 1937 20 -1917 343.0 2350 610 -1740 320.0 2359 544 -1815 225.0 2330 317 -2013 227.0 2115 534 -1581 319.0 2245 12 -2233 87.0 2254 655 -1599 301.0 2235 26 -2209 111.0 2231 16 -2215 105.0 2113 511 -1602 298.0 1600 13 -1587 313.0 2231 1 -2230 150.0 2245 615 -1630 270.0 1845 5 -1840 200.0 1940 151 -1789 251.0 2122 522 -1600 300.0 2141 59 -2082 198.0 2155 546 -1609 291.0 2305 648 -1657 283.0 1934 102 -1832 208.0 1444 629 -815 585.0 2015 706 -1309 411.0 2220 654 -1566 334.0 2045 500 -1545 315.0 2134 530 -1604 296.0 1955 55 -1900 180.0 1925 25 -1900 180.0 2124 550 -1574 326.0 2359 352 -2007 233.0 2202 54 -2148 172.0 2250 5 -2245 135.0 2215 545 -1670 270.0 1945 25 -1920 160.0 2145 25 -2120 160.0 2209 555 -1654 286.0 2155 50 -2105 115.0 2205 45 -2160 100.0 1925 50 -1875 205.0 2230 35 -2195 125.0 2320 15 -2305 55.0 2255 23 -2232 88.0 2330 20 -2310 50.0 1850 10 -1840 200.0 2115 16 -2099 241.0 2005 20 -1985 195.0 2155 45 -2110 170.0 2315 12 -2303 57.0 2230 51 -2179 141.0 2330 518 -1812 228.0 2259 26 -2233 87.0 2150 43 -2107 173.0 2100 527 -1573 327.0 2359 810 -1549 311.0 2115 513 -1602 298.0 2100 527 -1573 327.0 2232 2 -2230 90.0 2210 600 -1610 290.0 2145 555 -1590 310.0 2220 18 -2202 118.0 2225 3 -2222 98.0 1825 55 -1770 210.0 2239 32 -2207 113.0 1740 35 -1705 235.0 2220 49 -2171 149.0 2215 531 -1684 256.0 2215 15 -2200 120.0 1720 25 -1695 245.0 900 851 -49 51.0 1712 1705 -7 53.0 2205 620 -1585 315.0 1625 55 -1570 330.0 826 818 -8 52.0 2210 36 -2174 146.0 1805 55 -1750 230.0 2305 601 -1704 236.0 1845 29 -1816 224.0 2330 1 -2329 91.0 2235 604 -1631 269.0 2120 524 -1596 304.0 2128 531 -1597 303.0 2355 527 -1828 212.0 2320 750 -1570 330.0 2215 625 -1590 310.0 2205 503 -1702 238.0 2230 523 -1707 233.0 2200 444 -1756 224.0 2210 5 -2205 115.0 2000 55 -1945 175.0 1356 1348 -8 52.0 2155 20 -2135 145.0 2300 736 -1564 336.0 2230 147 -2083 377.0 2245 33 -2212 108.0 2250 4 -2246 74.0 2340 625 -1715 225.0 2340 751 -1589 311.0 2130 10 -2120 160.0 2240 121 -2119 221.0 2100 711 -1389 371.0 2250 732 -1518 342.0 1255 1248 -7 53.0 2220 37 -2183 137.0 2344 2339 -5 55.0 2338 2333 -5 55.0 2030 20 -2010 170.0 2212 2 -2210 290.0 1542 633 -909 531.0 2222 13 -2209 111.0 2214 55 -2159 221.0 2230 38 -2192 128.0 2356 549 -1807 233.0 2345 131 -2214 106.0 2355 555 -1800 240.0 2245 8 -2237 83.0 2220 625 -1595 305.0 2055 41 -2014 166.0 2330 707 -1623 277.0 2145 31 -2114 226.0 1925 17 -1908 232.0 1955 118 -1837 203.0 2235 604 -1631 269.0 1915 145 -1770 210.0 1951 1949 -2 58.0 2230 15 -2215 105.0 2035 502 -1533 327.0 2130 540 -1590 310.0 2259 29 -2230 90.0 2200 541 -1659 281.0 2225 430 -1795 245.0 2110 604 -1506 354.0 2145 207 -1938 202.0 2155 45 -2110 170.0 2145 44 -2101 119.0 2311 500 -1811 229.0 2359 344 -2015 225.0 2030 122 -1908 172.0 2359 247 -2112 168.0 2250 15 -2235 85.0 2350 754 -1596 304.0 2315 639 -1676 264.0 2229 4 -2225 95.0 2350 602 -1748 252.0 2209 600 -1609 291.0 2235 626 -1609 291.0 2320 620 -1700 240.0 2235 700 -1535 325.0 1657 45 -1612 288.0 1940 716 -1224 456.0 2140 20 -2120 100.0 2210 559 -1651 289.0 2336 2329 -7 53.0 2140 505 -1635 325.0 639 629 -10 50.0 1950 55 -1895 

##### 2.1.3.3 Weather

We have 30,528,602 records in our weather data set representing weather readings from a specific station (at a specific location) at a specific time.

In [ ]:
df_weather_full = spark.read.parquet(f"{data_BASE_DIR}parquet_weather_data/")
df_weather_full.createOrReplaceTempView("df_weather_full_tb")
df_weather_distinct = spark.sql(
  """
  select distinct 
  STATION,
  DATE,
  LATITUDE,
  LONGITUDE,
  ELEVATION,
  REPORT_TYPE,
 HourlyAltimeterSetting,
 HourlyDewPointTemperature,
 HourlyDryBulbTemperature,
 HourlyRelativeHumidity,
 HourlySkyConditions,
 HourlyStationPressure,
 HourlyVisibility,
 HourlyWetBulbTemperature,
 HourlyWindDirection,
 HourlyWindSpeed
  from df_weather_full_tb
  """)

df_weather_distinct.createOrReplaceTempView("df_weather_distinct")

__2.1.3.3.1 Weather - Sky Condition__


The code "CLR:00" indicates clear skies, and 4014537 of our 30528602 data points have this value (13%).  Propose creating a column "is_clear" with those.   4992829 of our data points have "OVC" for overcast in the description (16.4%).  Propose creating a "is_overcast" column with those.

___METAR Abbreviations___

OVC - Overcast
FEW - Few clouds

No "SN" or "TS" or "FZ" severe weather events in the dataset.

In [ ]:
# df_weather_distinct.createOrReplaceTempView("df_weather_distinct_temp")
df_weather_distinct.select("HourlySkyConditions").filter(col("HourlySkyConditions").like("%OVC%")).count()

Out[58]: 110468075

In [ ]:
%sql

select
 HourlySkyConditions, COUNT(*) as sum
from
  df_weather_distinct
WHERE HourlySkyConditions == "%FZ%"
GROUP BY HourlySkyConditions
ORDER BY sum DESC


HourlySkyConditions sum

__2.1.3.3.2 Weather - Report Type__

We have several reports for 1 station at 1 time based on the check below; therefore, we're going to take the most special events first and fill the missings with the average from all the other reports to capture the most significant weather events.

In [ ]:
%sql

with report_check as (
select 
station, date, 
count(report_type) as report_cnt
from df_weather_distinct
group by station, date)
select max(report_cnt) from report_check


max(report_cnt) 4

#### 2.2 Table description
a) df_flights: It contains a time series of the flight schedules from the year 2015 to 2021, including:
- Time period information:
  - Flight date and time
  - Day of the week/day of the month and year
- Flight operational data:
  - Scheduled departure/arrival time
  - Departure/arrival time
  - Departure/arrival delay measured in minutes, it is calculated as the difference between the scheduled and actual arrival/departure time.
  - Flight cancellation, in our case we will ignore the the canceled flights
- Metadata associated with the flight's origin and destination airports:
  - IATA airport code which is the airport location's unique 3 letter identifier
  - Airport state and city
- Carrier information
 
b) df_neighbor_stations: It provides metadata about weather stations with neighbor airports. It includes:
- Station unique identifier
- neighbor_call corresponds to the ICAO airport code which is defined by the International Civil Aviation Organization
- Neighbor airport name, state

c) df_weather_by_station: It contains a time series of weather information per weather station, including:
- Station unique identifier
- Time period information 
- Station metadata
- Weather metrics
- Weather date lag timestamp

d) df_iata_icao_codes: External resource that contains the mapping between the IATA and ICAO unique airport codes and the airport timezone (source: https://openflights.org/data.html)

#### 2.3 Table joins

a) df_neighbor_stations <> df_iata_icao_codes
- We will perform an inner join between the df_neighbor_stations and df_iata_icao_codes tables, using the df_neighbor_stations.neighbor_call and the df_iata_icao_codes.icao_code columns to enhance the df_neighbor_stations table with a new corresponding IATA airport code.

b) df_flights <> df_iata_icao_codes
- We will perform a left join between the df_flights and df_iata_icao_codes tables to obtain the timezone per airport by iata code. This will be used later to transform the timestamps to UTC.

b) df_neighbor_stations <> df_flights
- Both tables can be joined using the df_flights.ORIGIN and the df_flights.DEST columns, which contain the IATA airport code for the origin and destination airports, respectively,  with the 
df_neighbor_stations.iata_code that was created above. We will need to join the tables twice, once using the df_flights.ORIGIN and df_neighbor_stations.iata_code columns and the second join using df_flights.DEST and df_neighbor_stations.iata_code columns. We can perform inner joins between these two tables because we are only interested in flights with associated weather information.
- This relationship is many to many because multiple flights can be mapped to multiple neighbor stations.
- With the result of these joins, we will create a new table called df_flight_station that contains the origin/destination flight and airport information along with the associated weather station_id. This table can be stored in the Azure Blob storage.

c) df_flights_station <> df_weather_station
- We will join the df_flights_station and the df_weather_by_station tables using the df_flights_station.station_id and df_weather_by_station.station columns. To obtain the latest weather measurement two hours before the estimated flight departure, we created two columns that define the acceptable window of time (from 6 hours before the flight to 2 hours before the flight) that we should consider when selecting the latest weather measurement.

d) The final df_flights_weather_station contains the flight/weather data for the origin and destination airports. This table includes a column that contains the previous timestamp available for the weather station measurement timestamp (lag). This will be used to obtain the origin and destination lag weather measurements by merging df_flights_weather_station.ws_origin/dest_weather_lag matching the weather.DATE and station_id columns.

##### 2.3.1 Table Join Step 1 - Flight with Airpot Code

In [ ]:
# df_airlines_clean.createOrReplaceTempView("df_airlines_clean_temp")
df_iata_icao_codes.createOrReplaceTempView("df_iata_icao_codes_temp")

df_flights_with_timezone = spark.sql(
"""
select
df_airlines_clean_temp.*,
to_timestamp(CONCAT(FL_DATE, ' ', SUBSTR(CRS_DEP_TIME, 0, LEN(CRS_DEP_TIME) - 2), ':', SUBSTR(CRS_DEP_TIME, - 2))) as CRS_DEP_TIMESTAMP,
icode1.tz_time_zone as tz_time_zone_origin,
icode2.tz_time_zone as tz_time_zone_dest
from
df_airlines_clean_temp
left join df_iata_icao_codes_temp icode1 on df_airlines_clean_temp.ORIGIN = icode1.IATA
left join df_iata_icao_codes_temp icode2 on df_airlines_clean_temp.DEST = icode2.IATA
where icode1.tz_time_zone is not null  and icode1.tz_time_zone not like '%\\N%'
""")

##### 2.3.2 Table Join Step 2 - Flight Time Adjustments
1. Convert to UTC
2. Create Time Interval

In [ ]:
#Transform CRS departure timestamp to UTC
df_flights_with_timezone = df_flights_with_timezone.withColumn("CRS_DEP_TIMESTAMP", to_utc_timestamp("CRS_DEP_TIMESTAMP", col("tz_time_zone_origin")))
# df_flights_with_timezone.display()


In [ ]:
#Transform CRS flight departure time to timestamp and create flight departure and arrival timestamps 2 hours and 6 hours before the scheduled departure time
df_flights_with_timezone.createOrReplaceTempView("df_flights_with_timezone_temp")
df_flights = spark.sql(
"""
select
 CRS_DEP_TIMESTAMP - INTERVAL 6 HOURS as CRS_DEP_TIMESTAMP_START,
 CRS_DEP_TIMESTAMP - INTERVAL 2 HOURS as CRS_DEP_TIMESTAMP_END,
 df_flights_with_timezone_temp.*
from
 df_flights_with_timezone_temp
""")
#df_flights.select([count(when( col(c).isNull(), c)).alias(c) for c in df_flights.columns]).display()


# Join neighbor_stations and flights data 
df_neighbor_stations.createOrReplaceTempView("df_neighbor_stations_temp")
df_flights.createOrReplaceTempView("df_flights_temp")

df_flights_station = spark.sql(
"""
select
ns1.station_id as station_id_origin,
ns2.station_id as station_id_dest,
df_flights_temp.*
from
df_flights_temp
left join df_neighbor_stations_temp ns1 on df_flights_temp.ORIGIN = ns1.IATA
left join df_neighbor_stations_temp ns2 on df_flights_temp.DEST = ns2.IATA
""")
# df_flights_station.display()

df_flights_station.createOrReplaceTempView("df_flights_station_temp")

##### 2.3.3 Table Join Step 3 - Weather Data Adjustments
1. Convert to UTC
2. Report Type Dedup

In [ ]:
df_weather_adj_tz = spark.sql("""
select t1.station, 
case when (t2.tz_time_zone not like '%\\N%' and t2.tz_time_zone is not null)
  then coalesce(to_utc_timestamp(cast(t1.date as timestamp), t2.tz_time_zone), cast(t1.date as timestamp))
  else cast(t1.date as timestamp) end as date,
 t1.report_type,
 t1.latitude,
 t1.longitude,
 t1.elevation,
 t1.HourlyAltimeterSetting,
t1.HourlyDewPointTemperature,
t1.HourlyDryBulbTemperature,
t1.HourlyRelativeHumidity,
t1.HourlySkyConditions,
t1.HourlyStationPressure,
t1.HourlyVisibility,
t1.HourlyWetBulbTemperature,
t1.HourlyWindDirection,
t1.HourlyWindSpeed
from df_weather_distinct as t1
left join df_neighbor_stations_temp as t2
on cast(t1.station as float) = cast(t2.station_id as float)
""")

df_weather_adj_tz.createOrReplaceTempView("df_weather_adj_tz")

In [ ]:
report_type_rank = spark.sql("""
select 'CRN05' as report_type, 5 as rank
union all
select 'FM-12' as report_type, 4 as rank
union all
select 'FM-13' as report_type, 4 as rank
union all
select 'FM-14' as report_type, 4 as rank
union all
select 'FM-15' as report_type, 4 as rank
union all
select 'FM-16' as report_type, 1 as rank
union all
select 'MESOW' as report_type, 3 as rank
union all
select 'SAO' as report_type, 2 as rank
union all
select 'SAOSP' as report_type, 4 as rank
union all
select 'SHEF' as report_type, 5 as rank
union all
select 'SOD' as report_type, 6 as rank
union all
select 'SOM' as report_type, 6 as rank
union all
select 'SURF' as report_type, 5 as rank
union all
select 'SY-MT' as report_type, 6 as rank
""")

report_type_rank.createOrReplaceTempView("report_type_rank")

In [ ]:
df_weather_compress = spark.sql("""
with base as 
(select 
 t1.station,
 t1.date,
 t1.report_type,
 t1.latitude,
 t1.longitude,
 t1.elevation,
 cast(regexp_replace(t1.HourlyAltimeterSetting,'[^0-9.]+','') as float) as HourlyAltimeterSetting,
 cast(regexp_replace(t1.HourlyDewPointTemperature,'[^0-9.]+','') as float) as HourlyDewPointTemperature,
 cast(regexp_replace(t1.HourlyDryBulbTemperature,'[^0-9.]+','') as float) as HourlyDryBulbTemperature,
 cast(regexp_replace(t1.HourlyRelativeHumidity,'[^0-9.]+','') as float) as HourlyRelativeHumidity,
 case when t1.HourlySkyConditions like '%OVC%' then 1
      when t1.HourlySkyConditions like '%CLR%' then 0
 end as OCV_CLR,
 cast(regexp_replace(t1.HourlyStationPressure,'[^0-9.]+','') as float) as HourlyStationPressure,
 cast(regexp_replace(t1.HourlyVisibility,'[^0-9.]+','') as float) as HourlyVisibility,
 cast(regexp_replace(t1.HourlyWetBulbTemperature,'[^0-9.]+','') as float) as HourlyWetBulbTemperature,
 cast(regexp_replace(t1.HourlyWindDirection,'[^0-9.]+','') as float) as HourlyWindDirection,
 cast(regexp_replace(t1.HourlyWindSpeed,'[^0-9.]+','') as float) as HourlyWindSpeed,
 t2.rank, 
 min(t2.rank) over (partition by t1.station, t1.date) as top_rank
 from df_weather_adj_tz as t1
left join report_type_rank as t2
 on t1.report_type = t2.report_type
),
top_info as (
select 
  station, date, 
  max(report_type) as report_type,
  avg(LATITUDE) as LATITUDE,
  avg(LONGITUDE) as LONGITUDE,
  avg(ELEVATION) as ELEVATION,
  avg(HourlyAltimeterSetting) as HourlyAltimeterSetting,
  avg(HourlyDewPointTemperature) as HourlyDewPointTemperature,
  avg(HourlyDryBulbTemperature) as HourlyDryBulbTemperature,
  avg(HourlyRelativeHumidity) as HourlyRelativeHumidity,
  avg(OCV_CLR) as OCV_CLR,
  avg(HourlyStationPressure) as HourlyStationPressure,
  avg(HourlyVisibility) as HourlyVisibility,
  avg(HourlyWetBulbTemperature) as HourlyWetBulbTemperature,
  avg(HourlyWindDirection) as HourlyWindDirection,
  avg(HourlyWindSpeed) as HourlyWindSpeed
  from base where rank = top_rank
  group by station, date
),
other_info as (
select 
  station, date, 
  avg(LATITUDE) as LATITUDE,
  avg(LONGITUDE) as LONGITUDE,
  avg(ELEVATION) as ELEVATION,
  avg(HourlyAltimeterSetting) as HourlyAltimeterSetting,
  avg(HourlyDewPointTemperature) as HourlyDewPointTemperature,
  avg(HourlyDryBulbTemperature) as HourlyDryBulbTemperature,
  avg(HourlyRelativeHumidity) as HourlyRelativeHumidity,
  avg(OCV_CLR) as OCV_CLR,
  avg(HourlyStationPressure) as HourlyStationPressure,
  avg(HourlyVisibility) as HourlyVisibility,
  avg(HourlyWetBulbTemperature) as HourlyWetBulbTemperature,
  avg(HourlyWindDirection) as HourlyWindDirection,
  avg(HourlyWindSpeed) as HourlyWindSpeed
  from base where rank != top_rank
  group by station, date
),
info_keep as (
select distinct 
  t1.station, t1.date, 
  t1.REPORT_TYPE, 
  coalesce(t1.LATITUDE, t2.LATITUDE) as LATITUDE,
  coalesce(t1.LONGITUDE, t2.LONGITUDE) as LONGITUDE,
  coalesce(t1.ELEVATION, t2.ELEVATION) as ELEVATION,
  coalesce(t1.HourlyAltimeterSetting, t2.HourlyAltimeterSetting) as HourlyAltimeterSetting,
  coalesce(t1.HourlyDewPointTemperature, t2.HourlyDewPointTemperature) as HourlyDewPointTemperature,
  coalesce(t1.HourlyDryBulbTemperature, t2.HourlyDryBulbTemperature) as HourlyDryBulbTemperature,
  coalesce(t1.HourlyRelativeHumidity, t2.HourlyRelativeHumidity) as HourlyRelativeHumidity,
  coalesce(t1.OCV_CLR, t2.OCV_CLR) as OCV_CLR,
  coalesce(t1.HourlyStationPressure, t2.HourlyStationPressure) as HourlyStationPressure,
  coalesce(t1.HourlyVisibility, t2.HourlyVisibility) as HourlyVisibility,
  coalesce(t1.HourlyWetBulbTemperature, t2.HourlyWetBulbTemperature) as HourlyWetBulbTemperature,
  coalesce(t1.HourlyWindDirection, t2.HourlyWindDirection) as HourlyWindDirection,
  coalesce(t1.HourlyWindSpeed, t2.HourlyWindSpeed) as HourlyWindSpeed
from top_info as t1
  left join other_info as t2
  on t1.station = t2.station
  and t1.date = t2.date
),
for_filter as (
select *,
 (case when HourlyAltimeterSetting is null then 1 else 0 end)+
 (case when HourlyDewPointTemperature is null then 1 else 0 end)+
 (case when HourlyDryBulbTemperature is null then 1 else 0 end)+
 (case when HourlyRelativeHumidity is null then 1 else 0 end)+
 (case when OCV_CLR is null then 1 else 0 end)+
 (case when HourlyStationPressure is null then 1 else 0 end)+
 (case when HourlyVisibility is null then 1 else 0 end)+
 (case when HourlyWetBulbTemperature is null then 1 else 0 end)+
 (case when HourlyWindDirection is null then 1 else 0 end)+
 (case when HourlyWindSpeed is null then 1 else 0 end) as null_cnt
from info_keep
)
select *,
    LAG(DATE, 1) OVER (
        PARTITION BY STATION, to_date(DATE)
        ORDER BY DATE
    ) as weather_lag
from for_filter
where null_cnt <= 3
""")

# df_weather_compress.write.parquet(f"{blob_url}/df_weather_to_join_full_v1")
# df_weather_to_join_full = spark.read.parquet(f"{blob_url}/df_weather_to_join_full_v1")
df_weather_compress.createOrReplaceTempView("df_weather_to_join_full_temp")

##### 2.3.4 Table Join Step 4 - Join Weather to Flight
1. Join Weather Info by Station for Departure and Arrival Separately
2. Take the Latest Weather Record within the Flight Time Interval
3. Combine the Departure and Arrival Data

In [ ]:
#Joining flights with weather dataset
# df_weather_to_join_full.createOrReplaceTempView("df_weather_to_join_full_temp")
# df_flights_station.createOrReplaceTempView("df_flights_station_temp")
df_flights_weather_station_origin =  spark.sql(
"""
select
df_flights_station_temp.*,
ws_origin.DATE as ws_origin_DATE,
ws_origin.LATITUDE as ws_origin_LATITUDE,
ws_origin.LONGITUDE as ws_origin_LONGITUDE,
ws_origin.ELEVATION as ws_origin_ELEVATION,
ws_origin.HourlyAltimeterSetting as ws_origin_HourlyAltimeterSetting,
ws_origin.HourlyDewPointTemperature as ws_origin_HourlyDewPointTemperature,
ws_origin.HourlyDryBulbTemperature as ws_origin_HourlyDryBulbTemperature,
ws_origin.HourlyRelativeHumidity as ws_origin_HourlyRelativeHumidity,
ws_origin.OCV_CLR as ws_origin_OCV_CLR,
ws_origin.HourlyStationPressure as ws_origin_HourlyStationPressure,
ws_origin.HourlyVisibility as ws_origin_HourlyVisibility,
ws_origin.HourlyWetBulbTemperature as ws_origin_HourlyWetBulbTemperature,
ws_origin.HourlyWindDirection as ws_origin_HourlyWindDirection,
ws_origin.HourlyWindSpeed as ws_origin_HourlyWindSpeed,
ws_origin.weather_lag as ws_origin_weather_lag
from 
df_flights_station_temp
left join df_weather_to_join_full_temp ws_origin on df_flights_station_temp.station_id_origin = ws_origin.station 
and (ws_origin.DATE BETWEEN df_flights_station_temp.CRS_DEP_TIMESTAMP_START AND df_flights_station_temp.CRS_DEP_TIMESTAMP_END)
""")



In [ ]:
df_flights_weather_station_origin.createOrReplaceTempView("df_flights_weather_station_origin_temp")
# Aggregate data to obtain only the latest observation in the timeframe
df_flights_weather_station_origin_agg = spark.sql(
"""
SELECT
   t1.*
FROM
    df_flights_weather_station_origin_temp as t1
where
  t1.ws_origin_DATE = (select max(s1.ws_origin_DATE) from df_flights_weather_station_origin_temp as s1
                      where t1.station_id_origin = s1.station_id_origin
                      and t1.CRS_DEP_TIMESTAMP = s1.CRS_DEP_TIMESTAMP
                      and t1.TAIL_NUM = s1.TAIL_NUM
                      and t1.OP_CARRIER_FL_NUM = s1.OP_CARRIER_FL_NUM)

""")

In [ ]:
df_flights_weather_station_dest =  spark.sql(
"""
select
df_flights_station_temp.station_id_dest,
df_flights_station_temp.CRS_DEP_TIMESTAMP,
df_flights_station_temp.TAIL_NUM,
df_flights_station_temp.OP_CARRIER_FL_NUM,
ws.DATE as ws_dest_DATE,
ws.LATITUDE as ws_dest_LATITUDE,
ws.LONGITUDE as ws_dest_LONGITUDE,
ws.ELEVATION as ws_dest_ELEVATION,
ws.HourlyAltimeterSetting as ws_dest_HourlyAltimeterSetting,
ws.HourlyDewPointTemperature as ws_dest_HourlyDewPointTemperature,
ws.HourlyDryBulbTemperature as ws_dest_HourlyDryBulbTemperature,
ws.HourlyRelativeHumidity as ws_dest_HourlyRelativeHumidity,
ws.OCV_CLR as ws_dest_OCV_CLR,
ws.HourlyStationPressure as ws_dest_HourlyStationPressure,
ws.HourlyVisibility as ws_dest_HourlyVisibility,
ws.HourlyWetBulbTemperature as ws_dest_HourlyWetBulbTemperature,
ws.HourlyWindDirection as ws_dest_HourlyWindDirection,
ws.HourlyWindSpeed as ws_dest_HourlyWindSpeed,
ws.weather_lag as ws_dest_weather_lag
from 
df_flights_station_temp
left join df_weather_to_join_full_temp ws on df_flights_station_temp.station_id_dest = ws.station 
and (ws.DATE BETWEEN df_flights_station_temp.CRS_DEP_TIMESTAMP_START AND df_flights_station_temp.CRS_DEP_TIMESTAMP_END)
""")


In [ ]:
df_flights_weather_station_dest.createOrReplaceTempView("df_flights_weather_station_dest_temp")
df_flights_weather_station_dest_agg = spark.sql(
"""
SELECT
   t1.*
FROM
    df_flights_weather_station_dest_temp as t1
where
  t1.ws_dest_DATE = (select max(s1.ws_dest_DATE) from df_flights_weather_station_dest_temp as s1
                      where t1.station_id_dest = s1.station_id_dest
                      and t1.CRS_DEP_TIMESTAMP = s1.CRS_DEP_TIMESTAMP
                      and t1.TAIL_NUM = s1.TAIL_NUM
                      and t1.OP_CARRIER_FL_NUM = s1.OP_CARRIER_FL_NUM)

""")


In [ ]:
##joining origin and destination flight, weather and station data
df_flights_weather_station_origin_agg.createOrReplaceTempView("df_flights_weather_station_origin_agg_temp")
df_flights_weather_station_dest_agg.createOrReplaceTempView("df_flights_weather_station_dest_agg_temp")
df_flights_weather_station_origin_dest= spark.sql(
"""
select 
fws1.*,
ws_dest_DATE,
ws_dest_LATITUDE,
ws_dest_LONGITUDE,
ws_dest_ELEVATION,
ws_dest_HourlyAltimeterSetting,
ws_dest_HourlyDewPointTemperature,
ws_dest_HourlyDryBulbTemperature,
ws_dest_HourlyRelativeHumidity,
ws_dest_OCV_CLR,
ws_dest_HourlyStationPressure,
ws_dest_HourlyVisibility,
ws_dest_HourlyWetBulbTemperature,
ws_dest_HourlyWindDirection,
ws_dest_HourlyWindSpeed,
ws_dest_weather_lag
from 
df_flights_weather_station_origin_agg_temp fws1
left join df_flights_weather_station_dest_agg_temp fws2
  on fws1.station_id_dest = fws2.station_id_dest
  and fws1.CRS_DEP_TIMESTAMP = fws2.CRS_DEP_TIMESTAMP
  and fws1.TAIL_NUM = fws2.TAIL_NUM
  and fws1.OP_CARRIER_FL_NUM = fws2.OP_CARRIER_FL_NUM
""")

# df_flights_weather_station_origin_dest.write.mode("overwrite").parquet(f"{blob_url}/flights_weather_station")

#####2.3.5 Table Join Data sizes
- Data sizes
  
  - Join Weather to Flight origin and destination
    - df_weather_compress: Cleaned/Deduped weather data 30,528,602 records
    - df_flights_station: Cleaned/Deduped flight data containing airport and station metadata 41,551,808 records
    
  - Join Origin and Destination including flight data and latest weather measurement 2 hours prior CRS departure time
    - df_flights_weather_station_dest_agg 41,551,808 records
    - df_flights_weather_station_origin_agg 41,551,808 records

- Total join time

  - It took a total of 3.38 hours to perform the final join

- Cluster specification

  - 1-4 Workers: 16-64 GB Memory, 4-16 Cores
  - 1 Driver: 16 GB Memory, 4 Cores
  - Runtime: 11.3.x-cpu-ml-scala2.12

#### 2.4 Data Cleaning and Validation

We are joining several complex data sets and we expect to find a number of issues that require data cleaning.  Some of the levels of different categorical variables may need to be merged, we will need to handle missing values, and normalize other numeric quantities.

##### 2.4.1 NULL Values

We will drop all columns with more than 50% NULL values, which are identified through data summarization.

##### 2.4.2 Outliers

We will carefully report any outlier data that we believe should be excluded, and only exclude it after a thorough investigation.

##### 2.4.3 Normalization and Scaling

We will normalize variables should be treated numerically like weather information.

##### 2.4.4 Class Imbalance

We have about a 5:1 imbalance between our "no delay" and "delay" classes.  For some of our modelling work, we may need to boost the number of training examples from our "delay" category to build a predictive model.

#### 2.5 EDA - Post Join

In [ ]:
df_flights_weather_station_origin_dest = spark.read.parquet(f"{blob_url}/flights_weather_station")
df_flights_weather_station_origin_dest.display()

station_id_origin station_id_dest CRS_DEP_TIMESTAMP_START CRS_DEP_TIMESTAMP_END QUARTER MONTH DAY_OF_MONTH DAY_OF_WEEK FL_DATE OP_UNIQUE_CARRIER TAIL_NUM OP_CARRIER_FL_NUM ORIGIN ORIGIN_CITY_NAME ORIGIN_STATE_ABR ORIGIN_WAC DEST DEST_CITY_NAME DEST_STATE_ABR DEST_WAC CRS_DEP_TIME DEP_DEL15 DEP_DELAY_GROUP DEP_TIME_BLK TAXI_IN CRS_ARR_TIME CRS_ELAPSED_TIME DISTANCE DISTANCE_GROUP YEAR CRS_DEP_TIMESTAMP tz_time_zone_origin tz_time_zone_dest ws_origin_DATE ws_origin_LATITUDE ws_origin_LONGITUDE ws_origin_ELEVATION ws_origin_HourlyAltimeterSetting ws_origin_HourlyDewPointTemperature ws_origin_HourlyDryBulbTemperature ws_origin_HourlyRelativeHumidity ws_origin_OCV_CLR ws_origin_HourlyStationPressure ws_origin_HourlyVisibility ws_origin_HourlyWetBulbTemperature ws_origin_HourlyWindDirection ws_origin_HourlyWindSpeed ws_origin_weather_lag ws_dest_DATE ws_dest_LATITUDE ws_dest_LONGITUDE ws_dest_ELEVATION ws_dest_HourlyAltimeterSetting ws_dest_HourlyDewPointTemperature ws_dest_HourlyDryBulbTemperature ws_dest_HourlyRelativeHumidity ws_dest_OCV_CLR ws_dest_HourlyStationPressure ws_dest_HourlyVisibility ws_dest_HourlyWetBulbTemperature ws_dest_HourlyWindDirection ws_dest_HourlyWindSpeed ws_dest_weather_lag 72219013874 null 2015-01-03T08:15:00.000+0000 2015-01-03T12:15:00.000+0000 1 1 3 6 2015-01-03 DL N826DN 878 ATL Atlanta, GA GA 34 SJU San Juan, PR PR 3 915 0.0 -1 0900-0959 3.0 1346 211.0 1547.0 7 2015 2015-01-03T14:15:00.000+0000 America/New_York America/Puerto_Rico 2015-01-03T12:00:00.000+0000 33.6301 -84.4418 307.8 null 48.0 49.0 97.0 null 29.1200008392334 0.5 48.0 90.0 10.0 2015-01-03T11:52:00.000+0000 null null null null null null null null null null null null null null null 74783012849 null 2015-01-05T22:30:00.000+0000 2015-01-06T02:30:00.000+0000 1 1 5 1 2015-01-05 B6 N328JB 1853 FLL Fort Lauderdale, FL FL 33 SJU San Juan, PR PR 3 2330 0.0 0 2300-2359 4.0 254 144.0 1046.0 5 2015 2015-01-06T04:30:00.000+0000 America/New_York America/Puerto_Rico 2015-01-06T02:06:00.000+0000 26.07875 -80.16217 3.4 30.290000915527344 69.0 74.0 85.0 null 30.280000686645508 10.0 71.0 50.0 9.0 2015-01-06T01:53:00.000+0000 null null null null null null null null null null null null null null null 72406093721 null 2015-01-08T07:40:00.000+0000 2015-01-08T11:40:00.000+0000 1 1 8 4 2015-01-08 WN N8325D 451 BWI Baltimore, MD MD 35 SJU San Juan, PR PR 3 840 0.0 0 0800-0859 2.0 1355 255.0 1565.0 7 2015 2015-01-08T13:40:00.000+0000 America/New_York America/Puerto_Rico 2015-01-08T10:54:00.000+0000 39.1733 -76.684 47.5 30.520000457763672 6.0 8.0 52.0 0.0 30.350000381469727 10.0 6.0 290.0 11.0 2015-01-08T09:54:00.000+0000 null null null null null null null null null null null null null null null 72508014740 null 2015-01-10T13:08:00.000+0000 2015-01-10T17:08:00.000+0000 1 1 10 6 2015-01-10 B6 N659JB 275 BDL Hartford, CT CT 11 SJU San Juan, PR PR 3 1408 1.0 1 1400-1459 4.0 1901 233.0 1666.0 7 2015 2015-01-10T19:08:00.000+0000 America/New_York America/Puerto_Rico 2015-01-10T16:51:00.000+0000 41.9375 -72.6819 53.3 30.389999389648438 0.0 20.0 41.0 null 30.190000534057617 10.0 15.0 300.0 9.0 2015-01-10T15:51:00.000+0000 null null null null null null null null null null null null null null null 72202012839 null 2015-01-12T06:05:00.000+0000 2015-01-12T10:05:00.000+0000 1 1 12 1 2015-01-12 AA N5EKAA 1506 MIA Miami, FL FL 33 SJU San Juan, PR PR 3 705 1.0 12 0700-0759 6.0 1035 150.0 1045.0 5 2015 2015-01-12T12:05:00.000+0000 America/New_York America/Puerto_Rico 2015-01-12T10:03:00.000+0000 25.7881 -80.3169 8.8 30.15999984741211 70.0 76.0 82.0 1.0 30.1299991607666 2.5 72.0 100.0 17.0 2015-01-12T09:53:00.000+0000 null null null null null null null null null null null null null null null 72211012842 null 2015-01-14T07:23:00.000+0000 2015-01-14T11:23:00.000+0000 1 1 14 3 2015-01-14 B6 N658JB 651 TPA Tampa, FL FL 33 SJU San Juan, PR PR 3 823 0.0 -1 0800-0859 4.0 1212 169.0 1237.0 5 2015 2015-01-14T13:23:00.000+0000 America/New_York America/Puerto_Rico 2015-01-14T10:53:00.000+0000

In [ ]:
df_flights_weather_station_origin_dest.createOrReplaceTempView("joined_table_temp")

___Delay vs Non-Delay by State___
<br>`The plot shows the number of delayed and non-delay flights by departure state`
<br>No obvious difference between delay vs non-delay flights.


___Delay vs Non-Delay by Dew Point Temp___
<br>`The plot shows the number of delayed and non-delay flights by Dew Point Temperature`
<br>The relative steep increase from 10 to 30 could be observed from non-delay which is not that obvious for delayed flights.


___Delay vs Non-Delay by Dry Buld Temp___
<br>`The plot shows the number of delayed and non-delay flights by Dry Bulb Temperature`
<br>The trends are not significantly differ between delayed vs non-delayed flights.


___Delay vs Non-Delay by Relative Humidity___
<br>`The plot shows the number of delayed and non-delay flights by Relative Humidity`
<br>The overal trends are similar across delayed vs non-delayed flights, but the relative changes are not always consistent.


___Delay vs Non-Delay by Pressure___
<br>`The plot shows the number of delayed and non-delay flights by Station Pressure`
<br>The trends are not significantly differ between delayed vs non-delayed flights.

In [ ]:
%sql

select 
year, quarter, month,
origin_state_abr, 
ws_origin_HourlyDewPointTemperature,
ws_origin_HourlyDryBulbTemperature,
ws_origin_HourlyRelativeHumidity,
ws_origin_HourlyStationPressure,
ws_origin_HourlyVisibility,
ws_origin_HourlyWetBulbTemperature,
ws_origin_HourlyWindSpeed,
sum(dep_del15) as delay_cnt,
sum(1-dep_del15) as non_deplay_cnt
from joined_table_temp
group by year, quarter, month,
origin_state_abr, 
ws_origin_HourlyDewPointTemperature,
ws_origin_HourlyDryBulbTemperature,
ws_origin_HourlyRelativeHumidity,
ws_origin_HourlyStationPressure,
ws_origin_HourlyVisibility,
ws_origin_HourlyWetBulbTemperature,
ws_origin_HourlyWindSpeed;

year quarter month origin_state_abr ws_origin_HourlyDewPointTemperature ws_origin_HourlyDryBulbTemperature ws_origin_HourlyRelativeHumidity ws_origin_HourlyStationPressure ws_origin_HourlyVisibility ws_origin_HourlyWetBulbTemperature ws_origin_HourlyWindSpeed delay_cnt non_deplay_cnt 2018 3 8 FL 75.0 83.0 77.0 29.979999542236328 10.0 77.0 15.0 3.0 21.0 2016 2 6 AZ 44.0 109.0 11.0 28.600000381469727 10.0 69.0 10.0 4.0 15.0 2016 3 9 TX 68.0 82.0 63.0 29.489999771118164 10.0 73.0 13.0 2.0 20.0 2019 2 6 HI 70.0 77.0 79.0 29.93000030517578 10.0 72.0 5.0 1.0 4.0 2019 1 1 GA 29.0 43.0 58.0 29.15999984741211 9.9399995803833 37.0 16.0 4.0 42.0 2015 2 5 PA 62.0 65.0 90.0 30.229999542236328 4.0 63.0 8.0 1.0 12.0 2015 3 8 NV 48.0 97.0 19.0 27.489999771118164 10.0 67.0 7.0 6.0 17.0 2017 3 7 IL 54.0 84.0 36.0 29.350000381469727 10.0 66.0 3.0 3.0 5.0 2019 4 10 OH 34.0 41.0 76.0 29.110000610351562 10.0 38.0 6.0 3.0 15.0 2016 2 6 MI 54.0 66.0 65.0 29.350000381469727 10.0 59.0 8.0 3.0 33.0 2015 1 1 CO 21.0 27.0 78.0 24.75 9.9399995803833 25.0 7.0 0.0 7.0 2016 3 7 MD 62.0 89.0 41.0 29.600000381469727 10.0 71.0 11.0 6.0 10.0 2018 3 9 MO 66.0 92.0 43.0 29.31999969482422 10.0 74.0 11.0 4.0 17.0 2015 1 3 AR 47.0 56.0 72.0 28.530000686645508 10.0 51.0 8.0 0.0 2.0 2015 4 12 LA 34.0 47.0 61.0 30.329999923706055 10.0 41.0 14.0 1.0 7.0 2016 1 1 AZ 38.0 42.0 85.0 29.030000686645508 10.0 40.0 0.0 7.0 28.0 2017 2 5 OK 60.0 81.0 49.0 29.270000457763672 10.0 68.0 14.0 1.0 2.0 2017 3 7 SD 55.0 59.0 87.0 28.5 10.0 57.0 8.0 0.0 4.0 2019 2 5 AL 64.0 68.0 87.0 29.639999389648438 10.0 66.0 7.0 0.0 2.0 2019 2 6 OR 55.0 73.0 53.0 29.90999984741211 10.0 62.0 7.0 0.0 5.0 2015 2 6 SD 56.0 66.0 70.0 26.68000030517578 9.9399995803833 60.0 5.0 0.0 1.0 2019 2 4 ME 27.0 38.0 65.0 29.700000762939453 10.0 34.0 13.0 0.0 1.0 2015 2 6 AK 41.0 54.0 62.0 29.690000534057617 8.0 48.0 11.0 1.0 0.0 2019 2 5 AK 31.0 33.0 92.0 30.059999465942383 10.0 32.0 14.0 0.0 1.0 2019 3 9 TX 70.0 89.0 53.0 29.299999237060547 9.9399995803833 76.0 11.0 16.0 58.0 2015 4 10 NY 46.0 59.0 62.0 29.68000030517578 9.9399995803833 52.0 9.0 0.0 19.0 2017 1 2 FL 65.0 68.0 90.0 29.969999313354492 10.0 66.0 5.0 1.0 28.0 2019 2 5 NC 64.0 66.0 93.0 29.329999923706055 10.0 65.0 14.0 7.0 13.0 2019 3 8 VA 66.0 81.0 61.0 29.799999237060547 9.9399995803833 71.0 5.0 6.0 21.0 2019 4 10 NY 47.0 52.0 83.0 29.989999771118164 10.0 49.0 6.0 2.0 15.0 2019 4 12 CA 12.0 51.0 21.0 29.729999542236328 10.0 37.0 3.0 0.0 26.0 2015 1 2 VA 10.0 15.0 80.0 29.549999237060547 6.0 14.0 5.0 1.0 3.0 2015 1 2 TX 23.0 26.0 88.0 29.8700008392334 6.0 25.0 16.0 2.0 0.0 2015 2 5 OH 66.0 75.0 74.0 29.239999771118164 10.0 69.0 10.0 0.0 1.0 2015 4 10 IL 42.0 45.0 90.0 29.479999542236328 10.0 44.0 0.0 3.0 69.0 2016 2 5 NC 64.0 66.0 93.0 29.25 10.0 65.0 6.0 7.0 5.0 2016 4 12 FL 72.0 80.0 76.0 30.170000076293945 9.9399995803833 75.0 8.0 12.0 1.0 2017 3 7 PA 64.0 84.0 51.0 29.8799991607666 9.9399995803833 71.0 7.0 1.0 5.0 2017 4 12 TX 50.0 58.0 75.0 29.420000076293945 10.0 54.0 9.0 3.0 32.0 2018 1 2 NJ 47.0 59.0 64.0 29.600000381469727 10.0 53.0 13.0 6.0 18.0 2018 1 3 MO 48.0 51.0 89.0 28.559999465942383 5.590000152587891 49.0 17.0 0.0 5.0 2018 1 3 FL 45.0 54.0 72.0 30.15999984741211 10.0 50.0 6.0 0.0 18.0 2018 3 8 NJ 66.0 76.0 72.0 30.209999084472656 10.0 70.0 13.0 3.0 21.0 2018 3 9 NJ 71.0 73.0 94.0 30.0 10.0 72.0 7.0 0.0 34.0 2018 4 12 GA 55.0 58.0 90.0 28.670000076293945 1.0 56.0 8.0 6.0 21.0 2019 2 4 FL 54.0 81.0 39.0 29.950000762939453 10.0 65.0 11.0 4.0 11.0 2019 2 5 OH 39.0 46.0 77.0 29.100000381469727 10.0 43.0 11.0 1.0 12.0 2015 4 11 IL 49.0 53.0 86.0 29.510000228881836 10.0 51.0 17.0 0.0 26.0 2019 2 6 IL 55.0 66.0 68.0 29.100000381469727 9.9399995803833 60.0 13.0 11.0 41.0 2017 4 12 GA 30.0 37.0 76.0 28.899999618530273 9.9399995803833 34.0 10.0 18.0 38.0 2018 1 2 NC 52.0 67.0 59.0 29.229999542236328 10.0 58.0 5.0 12.0 64.0 2015 2 4 GA 48.0 66.0 52.0 28.760000228881836 10.0 56.0 9.0 2.0 49.0 2019 3 8 GA 67.0 89.0 48.0 28.8700008392334 10.0 74.0 

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

#### 2.6 Feature Dictionary Post Join
|Feature                                |  Description                                                                      |
|---------------------------------------|-----------------------------------------------------------------------------------|      
|  **Flight / Airport**                                                                                                     |
|  QUARTER                              |  Flight Quarter in a calendar year (1-4)                                          |    
|  MONTH                                |  Flight Month                                                                     |
|  DAY_OF_MONTH                         |  Flight Day of the month                                                          |
|  DAY_OF_WEEK  Flight                  |  Flight Day of week                                                               |
|  FL_DATE                              |  Flight date (YYYY-MM-DD)                                                         |
|  OP_UNIQUE_CARRIER                    |  Flight Carrier unique identifier                                                 |
|  TAIL_NUM                             |  Tail Number Flight tracker                                                       |
|  OP_CARRIER_FL_NUM                    |  Carrier-Flight identifier                                                        |
|  ORIGIN                               |  Origin Airport IATA code                                                         |                          
|  ORIGIN_CITY_NAME                     |  Origin Airport City Name                                                         | 
|  ORIGIN_STATE_ABR                     |  Origin Airport State abbreviation                                                | 
|  ORIGIN_WAC                           |  Origin Airport World Area code                                                   |
|  DEST                                 |  Destination Airport IATA code                                                    |
|  DEST_CITY_NAME                       |  Destination Airport City Name                                                    |
|  DEST_STATE_ABR                       |  Destination Airport State abbreviation                                           |
|  DEST_WAC                             |  Destination Airport World Area code                                              |
|  DEP_DEL15                            |  Departure delay of 15 minutes or More                                            |
|  DEP_TIME_BLK                         |  CRS estimated Departure Time Block, Hourly Intervals                             |
|  CRS_ARR_TIME                         |  CRS estimated Arrival Time                                                       |
|  CRS_ELAPSED_TIME                     |  CRS estimated flight elapsed time                                                |
|  DISTANCE                             |  Distance between airports                                                        |
|  DISTANCE_GROUP                       |  Distance Intervals, every 250 Miles, for Flight Segment                          | 
|  YEAR                                 |  Flight year                                                                      |
|  CRS_DEP_TIMESTAMP                    |  CRS_DEP_TIMESTAMP, result of the transformation of FL_DATE and CRS_DEP_TIME      |
|  **Weather**                                                                                                              |
|  station_id_origin                    |  Origin station unique identifier                                                 |
|  station_id_dest                      |  Destination station unique identifier                                            |
|  ws_origin_DATE                       |  Origin Weather station origin measurement timestamp                              |
|  ws_origin_LATITUDE                   |  Origin Weather station Latitude                                                  |
|  ws_origin_LONGITUDE                  |  Origin Weather station Longitude                                                 |
|  ws_origin_ELEVATION                  |  Origin Weather station elevation                                                 |
|  ws_origin_HourlyAltimeterSetting     |  Origin Weather station Hourly Altimeter measurement                              |
|  ws_origin_HourlyDewPointTemperature  |  Origin Weather station Hourly Dew Point Temperature measurement                  |
|  ws_origin_HourlyDryBulbTemperature   |  Origin Weather station Hourly Dry Bulb Temperature measurement                   |
|  ws_origin_HourlyRelativeHumidity     |  Origin Weather station Hourly Relative Humidity measurement                      |
|  ws_origin_OCV_CLR                    |  Origin Weather station Hourly Sky Code, OCV(overcloud) = 1 CLR(clear) = 0        |
|  ws_origin_HourlyStationPressure      |  Origin Weather station Hourly Station Pressure measurement                       |
|  ws_origin_HourlyVisibility           |  Origin Weather station Hourly Visibility measurement                             |
|  ws_origin_HourlyWetBulbTemperature   |  Origin Weather station Hourly Wet Bulb Temperature measurement                   |
|  ws_origin_HourlyWindDirection        |  Origin Weather station Hourly Wind Direction measurement                         |
|  ws_origin_HourlyWindSpeed            |  Origin Weather station Hourly Wind Speed measurement                             |
|  ws_origin_weather_lag                |  Origin Weather station prior measurement timestamp                               |
|  ws_dest_DATE                         |  Destination Weather station origin measurement timestamp                         |
|  ws_dest_LATITUDE                     |  Destination Weather station Latitude                                             |
|  ws_dest_LONGITUDE                    |  Destination Weather station Longitude                                            |
|  ws_dest_ELEVATION                    |  Destination Weather station elevation                                            |
|  ws_dest_HourlyAltimeterSetting       |  Destination Weather station Hourly Altimeter measurement                         |
|  ws_dest_HourlyDewPointTemperature    |  Destination Weather station Hourly Dew Point Temperature measurement             |
|  ws_dest_HourlyDryBulbTemperature     |  Destination Weather station Hourly Dry Bulb Temperature measurement              |
|  ws_dest_HourlyRelativeHumidity       |  Destination Weather station Hourly Relative Humidity measurement                 |
|  ws_dest_OCV_CLR                      |  Destination Weather station Hourly Sky Code, OCV(overcloud) = 1 CLR(clear) = 0   |
|  ws_dest_HourlyStationPressure        |  Destination Weather station Hourly Station Pressure measurement                  |
|  ws_dest_HourlyVisibility             |  Destination Weather station Hourly Visibility measurement                        |
|  ws_dest_HourlyWetBulbTemperature     |  Destination Weather station Hourly Wet Bulb Temperature measurement              |
|  ws_dest_HourlyWindDirection          |  Destination Weather station Hourly Wind Direction measurement                    |
|  ws_dest_HourlyWindSpeed              |  Destination Weather station Hourly Wind Speed measurement                        |
|  ws_dest_weather_lag                  |  Destination Weather station prior measurement timestamp                          |

### 3. Machine Learning Algorithms

#### 3.1 Logistic Regression

Our first ML algorithm will be Logistic Regression.  We believe that with a variety of regression variables available to us that we can explore the feature space and build a logistic regression model that uses the source and destination city, airline, weather, seasonality, recent delay history, and many other robust features described above to classify all flights into two categories -- those with >= 15 minute delay of departure, and those without.

##### 3.1.1 Example Simple Logistic Regression Model

$$logit(p_i) = \beta_0 + \beta_1 x_{1,i} + \beta_2 x_{2,i} + ... + \beta_n x_{n,i}$$

Where features will be things like "airline carrier", "source city", "destination city", "raining at source city", "raining at destination city", "is winter", "is thunderstorm", etc.

We have a large number of features to choose from, and so we will explore adding an L2 regularization term to avoid overfitting on large set of features.

#### 3.2 Random Forests

We will also look to fit a Random Forest model to this data set.  We believe a random forest will be useful because it is robust to inclusion of irrelevant features, and invariant under scaling and transformation of many of the feature values.

### 4. Machine Learning Pipelines

#### 4.1 Pipeline description

a) Data Engineering:
- Step 1	
  - Ingest provided parquet files containing flights, weather, and station information and create corresponding dataframes.
  - Ingest external source airport code data in parquet format and create dataframe.
  - Perform EDA.
  - Data Cleaning.
  - Create a final dataframe with the result of the joined dataframes and store it in Azure Blob Storage in parquet format.

- Step 2
  - Ingest final_dataset from  Azure Blob Storage.
  - Select features
  - Split the data into Train, Validation, and Test dataset.
  - Perform normalization on the Train, Validation, and Test dataset using the Train dataset.
  - Store the normalized Train, Validation, and Test dataset Azure Blob Storage in parquet format.

b) Model Training
- Ingest the Train and Validation datasets from  Azure Blob Storage
- Build baseline model
- Build and Train the classification model
- Perform cross-validation and hyperparameter tuning

c) Model Evaluation
- Ingest the Test dataset from Azure Blob Storage
- Run the trained model on the Test dataset
- Evaluate model
- Store predictions in Azure Blob Storage

#### 4.2 Pipeline Block Diagram
<img src="https://github.com/carla-cortez/261/blob/main/ml_diagram.png?raw=true>" width=75%>

#### 4.3 Data Split Plan

The time serires data from 2015 to 2021 will be split in Train, Validation and Test datasets. We will remove the data from the year 2020 as it is considered an outlier.
- Train dataset: 2015 - 2018
- Validation dataset: 2019
- Test dataset: 2021

For cross-validation:
We will use a rolling window cross validation technique for time series. Having a fold for each year of data. For example:
- train on data from 2015 to predict 2016.  
- train on data from 2016 to predict 2017.
- train on data from 2017 to predict 2018.

In [ ]:
#split into train, validation, and test sets
#FYI wait till final join is done
#X and Y
df_train_X = flights_weather_station.where("YEAR BETWEEN '2015' AND '2018'").drop(col('DEP_DEL15'))
df_validation_X = flights_weather_station.where("YEAR == '2019'").drop(col('DEP_DEL15'))
df_test_X = flights_weather_station.where("YEAR == '2021'").drop(col('DEP_DEL15'))

In [ ]:
df_train_Y = flights_weather_station.where("YEAR BETWEEN '2015' AND '2018'").select(col('DEP_DEL15'))
df_validation_Y = flights_weather_station.where("YEAR == '2019'").select(col('DEP_DEL15'))
df_test_Y =flights_weather_station.where("YEAR == '2021'").select(col('DEP_DEL15'))

In [ ]:
# cross validation

#train on data from 2015 to predict 2016
df_train_2015_X = flights_weather_station.where("YEAR == '2015'").drop(col('DEP_DEL15'))
df_train_2015_Y = flights_weather_station.where("YEAR == '2015'").select(col('DEP_DEL15'))

# 2016 actuals
df_2016_Y_actual = flights_weather_station.where("YEAR == '2016'").select(col('DEP_DEL15'))

#train on data from 2015 to 2016 to predict 2017
df_train_2015_to_2016_X = flights_weather_station.where("YEAR BETWEEN '2015' AND '2016'").drop(col('DEP_DEL15'))
df_train_2015_to_2016_Y = flights_weather_station.where("YEAR BETWEEN '2015' AND '2016'").select(col('DEP_DEL15'))

# 2017 actuals
df_2017_Y_actual = flights_weather_station.where("YEAR == '2017'").select(col('DEP_DEL15'))

#train on data from 2015 to 2017 to predict 2018
df_train_2015_to_2017_X = flights_weather_station.where("YEAR BETWEEN '2015' AND '2017'").drop(col('DEP_DEL15'))
df_train_2015_to_2017_Y = dflights_weather_station.where("YEAR BETWEEN '2015' AND '2017'").select(col('DEP_DEL15'))

df_2018_Y_actual = flights_weather_station,where("YEAR == '2018'").select(col('DEP_DEL15'))

#### 4.4 Loss Function

We will aim to minimize our Log Loss function:

$$ Log Loss = \sum_{(x,y) \in D} -y log(y') - (1-y)log(1-y')$$

#### 4.5 Experiments
We are coming at this problem from the business perspective of our ability to predict a delay so that we can minimize the impact of our effected passengers.  For this reason, the cost of a False Negative is more important to us than the cost of a False Positive.

##### 4.5.1 Baseline

Across our entire data set, we see that 17.4% of flights are delayed more than 15 minutes from their scheduled departure.  We set as our baseline predictor a model that always predicts a delay.  We compute the precision and recall of this "Always Predict Delay" model, and will measure our logistic regression and other models against this baseline.

In [ ]:
%sql

select
--year, month,
-- concat(cast(year as string), case when month >= 10 then cast(month as string) else concat('0',cast(month as string)) end) as year_month,
-- sum(case when dep_delay>15 then 1 else 0 end) as delay_cnt,
sum(case when dep_delay>15 then 1 else 0 end)/count(dep_delay) as delay_pct
from df_airlines_full_tb
where year <= 2018
and cancelled = 0;

delay_pct 0.174117145151199

In [ ]:
%sql

select
  mean(DEP_DEL15)
from df_airlines_full_tb
where year == 2019
and cancelled = 0;

mean(DEP_DEL15) 0.18657226621465403

For a given observed delay percentage D, assumed to be 17.4% here, these metrics for an "Always predict delay" strategy are computed as

Precision = $$\frac{17.4}{17.4 + (100-17.4)} = \frac{17.4}{100} = .174$$
Recall = $$\frac{17.4}{17.4 + 0} = 1$$
F1 = $$\frac{2 * 17.4}{2 * 17.4 + (100-17.4) + 0} = \frac{34.8}{117.4} = .296$$

We intend to update these metrics with our logistic regression model and other models in later phases of the project.

| Model | Precision | Recall | F1 |
|---|---|---|---|
| Always predict delay | .174 | 1.0 | .296 |
| Never predict delay  | 0 | 0 | 0 |
| Logistic Regression #1 | TBD | TBD | TBD |

##### 4.5.2 Logistic Regression

We use pyspark.mk.classification.LogisticRegression to build a model based on our training data to predict membership of the "delay" or "no delay" class.  To do this, we setup a pipeline that uses one-hot encoding to vectorize our categorical variables, and then use the VectorAssembler to prepare it for use by the LogisticRegression model.  The example code below starts with a simplified model using only the MONTH column.  Each month of the year has a delayed % between 5-25% and so we expect this to be roughly equivalent as just always assuming all flights are not delayed until we add additional features into our logistic regression model.

In [ ]:
#base model 

from pyspark.ml.classification import LogisticRegression
from pyspark.ml.feature import OneHotEncoder, StringIndexer, VectorAssembler
from pyspark.ml import Pipeline

#df_murray = df_flights_weather_station.where("YEAR BETWEEN '2015' AND '2018'").select([ c for c in df_flights_weather_station.columns if c in {"DEP_DEL15", "MONTH"}])
df_lr_train = df_flights_weather_station_origin_dest.where("YEAR BETWEEN '2015' AND '2016'").select([ c for c in df_flights_weather_station_origin_dest.columns if c in {"DEP_DEL15", "MONTH"}])
df_lr_train = df_lr_train.withColumnRenamed("DEP_DEL15", "label")
cols = df_lr_train.columns
print("Columns: ", df_lr_train.columns)

print("average label by month")
monthly_avg = df_lr_train.groupBy("MONTH").mean()
monthly_avg.show()
categoricalColumns = ['MONTH']
stages = []

for categoricalCol in categoricalColumns:
    stringIndexer = StringIndexer(inputCol = categoricalCol, outputCol = categoricalCol + 'Index')
    encoder = OneHotEncoder(inputCols=[stringIndexer.getOutputCol()], outputCols=[categoricalCol + "classVec"])
    stages += [stringIndexer, encoder]
    
assemblerInputs = [c + "classVec" for c in categoricalColumns] # + numericCols

assembler = VectorAssembler(inputCols=assemblerInputs, outputCol="features")
stages += [assembler]

pipeline = Pipeline(stages = stages)
pipelineModel = pipeline.fit(df_lr_train)
df = pipelineModel.transform(df_lr_train)

print("df columns: ", df.columns)
selectedCols = ['features'] + cols
df = df.select(selectedCols)
print("df columns: ", df.columns)
df.printSchema()


#params copied from spark manual - change later
lr = LogisticRegression(maxIter=10, regParam=0.3, elasticNetParam=0.8)

# Fit the model
lrModel = lr.fit(df)

#loss function


# training summary 
trainingSummary = lrModel.summary

print("lrmodel: ", lrModel)



Columns:  ['MONTH', 'label']
average label by month
+-----+----------+-------------------+
|MONTH|avg(MONTH)|         avg(label)|
+-----+----------+-------------------+
|   12|      12.0|0.21551831404677235|
|    1|       1.0|0.18120657490782496|
|    6|       6.0| 0.2211132952569639|
|    3|       3.0|0.18159546521729167|
|    5|       5.0| 0.1698913585334911|
|    9|       9.0| 0.1298328464427702|
|    4|       4.0|0.15536667325558412|
|    8|       8.0| 0.1978897665512522|
|    7|       7.0|0.22164795318712646|
|   10|      10.0|  0.129465040939234|
|   11|      11.0|0.14144587993528152|
|    2|       2.0|0.18498993227164562|
+-----+----------+-------------------+

df columns:  ['MONTH', 'label', 'MONTHIndex', 'MONTHclassVec', 'features']
df columns:  ['features', 'MONTH', 'label']
root
 |-- features: vector (nullable = true)
 |-- MONTH: integer (nullable = true)
 |-- label: double (nullable = true)

lrmodel:  LogisticRegressionModel: uid=LogisticRegression_08d84ea00e63, numClasses=

In [ ]:
df_lr_test = df_flights_weather_station_origin_dest.where("YEAR == '2017'").select([ c for c in df_flights_weather_station_origin_dest.columns if c in {"DEP_DEL15", "MONTH"}])
df_lr_test = df_lr_test.withColumnRenamed("DEP_DEL15", "label")
df_lr_test_x = df_lr_test.select(col("MONTH"))

dftest = pipelineModel.transform(df_lr_test)
predictions = lrModel.transform(dftest)
print(predictions.head(4))

precision = predictions.filter(predictions.label == predictions.prediction).count() / float(predictions.count())
print("Precision : ", precision)
print("Train rows: ", df_lr_train.count())
print("Test rows: ", df_lr_test.count())

[Row(MONTH=1, label=1.0, MONTHIndex=10.0, MONTHclassVec=SparseVector(11, {10: 1.0}), features=SparseVector(11, {10: 1.0}), rawPrediction=DenseVector([1.5301, -1.5301]), probability=DenseVector([0.822, 0.178]), prediction=0.0), Row(MONTH=1, label=0.0, MONTHIndex=10.0, MONTHclassVec=SparseVector(11, {10: 1.0}), features=SparseVector(11, {10: 1.0}), rawPrediction=DenseVector([1.5301, -1.5301]), probability=DenseVector([0.822, 0.178]), prediction=0.0), Row(MONTH=1, label=0.0, MONTHIndex=10.0, MONTHclassVec=SparseVector(11, {10: 1.0}), features=SparseVector(11, {10: 1.0}), rawPrediction=DenseVector([1.5301, -1.5301]), probability=DenseVector([0.822, 0.178]), prediction=0.0), Row(MONTH=1, label=0.0, MONTHIndex=10.0, MONTHclassVec=SparseVector(11, {10: 1.0}), features=SparseVector(11, {10: 1.0}), rawPrediction=DenseVector([1.5301, -1.5301]), probability=DenseVector([0.822, 0.178]), prediction=0.0)]
Accuracy :  0.818962582873523
Train rows:  11192253
Test rows:  5549571


In [ ]:
f1 = 2* predictions.filter(predictions.label == predictions.prediction).count() / (2*predictions.filter(predictions.label == predictions.prediction).count() + predictions.filter(predictions.label != predictions.prediction).count())
print(f1)
predictions.groupBy(["label", "prediction"]).count().show()
2*

0.9004722490878909
+-----+----------+-------+
|label|prediction|  count|
+-----+----------+-------+
| null|       0.0|      1|
|  1.0|       0.0|1004679|
|  0.0|       0.0|4544891|
+-----+----------+-------+



##### 4.5.3 Experiments Summary

|   | Baseline Experiment | Years | Input Features | Train | Valid | Test | Metrics - Precision | Metrics - Recall | Metrics - F1 |
|---|---|---|---|---|---|---|---|---|---|
| Baseline | Always Predict Delay | All | Nan | 14,272,964 | NA | 3,504,568 | 0.174 | 1.0 | 0.296 |
| Baseline | Always Predict NoDelay | All |Nan | 14,272,964 | NA | 3,504,568 |  |  | Undefined |
| Logistic Regression | Always Predict NoDelay | 2017 | Month | 11,192,253 | 5,549,571 |  | | | Undefined |


<br>

__Metrics__

Precision and recall are defined in terms of True positive (TP), False positive (FP), True Negative (TN), and False Negative (FN) classifications.

Precision = $$\frac{TP}{TP+FP}$$

Recall = $$\frac{TP}{TP+FN}$$

F1 = $$\frac{2TP}{2TP+FP+FN}$$

### 5. Project Timeline

<img src="https://github.com/carla-cortez/261/blob/main/gantt_diagram.png?raw=true>" width=75%>